File: 2021/5/part1.py
with open("input") as f:
    raw = f.readlines()
    data = []
    for line in raw:
        [start, end] = line.split(" -> ")
        [a, b] = start.split(",")
        [c, d] = end.split(",")
        data.append(((int(a), int(b)), (int(c), int(d))))

    counts = {}

    for (start, end) in data:
        (x1, y1) = start
        (x2, y2) = end
        if x1 == x2:
            for i in range(min(y1, y2), max(y1, y2) + 1):
                if (x1, i) in counts:
                    counts[(x1, i)] = counts[(x1, i)]  + 1
                else:
                    counts[(x1, i)] = 1
        if y1 == y2:
            for i in range(min(x1, x2), max(x1, x2) + 1):
                if (i, y1) in counts:
                    counts[(i, y1)] = counts[(i, y1)]  + 1
                else:
                    counts[(i, y1)] = 1
    total = 0
    for item in counts:
        if counts[item] >= 2:
            total += 1
    print(total)

File: 2021/14/part1.pl
use warnings;
use strict;
no warnings ('substr');

my $filename = 'input';

open(FH, '<', $filename) or die $!;

my %data = ();
my @lines = ();
while(<FH>) {
  my $item = $_;
  $item =~ s/\n//;
  push @lines, $item;
}
close(FH);
my $input = shift @lines;
$input =~ s/\s//;

shift @lines;

foreach (@lines) {
  my @x = split(" -> ", $_);
  $data{$x[0]} = $x[1];
}
my @range = (1..10);
for (@range) {
  my $new_input = '';

  while (length($input) > 1) {
    my $identifier = substr $input, 0, 2;
    my $target = $data{$identifier};
    $input = (substr $input, 1) || '';
    $new_input = $new_input . (substr $identifier, 0, 1) . $target;
  }
  $new_input = $new_input . $input;
  $input = $new_input;
}

my %counts = ();
foreach (split "", $input) {
  $counts{$_} = exists($counts{$_}) ? $counts{$_} + 1 : 1;
}
my $min = 5000;
my $max = 0;

foreach(keys %counts) {
  my $value = $counts{$_};
  if ($value > $max) {
    $max = $value;
  }
  if ($value < $min) {
    $min = $value;
  }
}

print $max - $min . "\n";
